{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy\n",
    "import pandas\n",
    "\n",
    "from aavomics import database\n",
    "from aavomics import aavomics\n",
    "import anndata\n",
    "import scanpy\n",
    "\n",
    "from plotly import offline as plotly\n",
    "from plotly import graph_objects\n",
    "from plotly.subplots import make_subplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ANNDATA_FILE_NAME = \"aavomics_mouse_cortex_2021.h5ad\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the anndata file\n",
    "\n",
    "adata = anndata.read_h5ad(os.path.join(database.DATA_PATH, ANNDATA_FILE_NAME))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out unlabeled cells\n",
    "labeled_cell_mask = ~pandas.isna(adata.obs[\"X_CCN202105070\"])\n",
    "adata = adata[labeled_cell_mask]\n",
    "\n",
    "# Filter out genes without any detected expression\n",
    "gene_mask = (adata.X.max(axis=0) > 0).toarray().flatten()\n",
    "adata = adata[:, gene_mask].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a mask identifying endothelial cells vs others\n",
    "endothelial_cell_mask = adata.obs[\"Cell Type\"].isin([\"Endothelial Cells\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the t-SNE coordinates\n",
    "\n",
    "coordinates = numpy.array([\n",
    "    adata.obs[\"X_CCN202105070\"],\n",
    "    adata.obs[\"Y_CCN202105070\"]\n",
    "]).transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a t-SNE plot coloring endothelial vs other cells\n",
    "\n",
    "other_cells_trace = graph_objects.Scatter(\n",
    "    x=coordinates[~endothelial_cell_mask, 0],\n",
    "    y=coordinates[~endothelial_cell_mask, 1],\n",
    "    mode=\"markers\",\n",
    "    marker={\n",
    "        \"size\": 2\n",
    "    },\n",
    "    name=\"Other Cells\"\n",
    ")\n",
    "\n",
    "endothelial_trace = graph_objects.Scatter(\n",
    "    x=coordinates[endothelial_cell_mask, 0],\n",
    "    y=coordinates[endothelial_cell_mask, 1],\n",
    "    mode=\"markers\",\n",
    "    marker={\n",
    "        \"size\": 2\n",
    "    },\n",
    "    name=\"Endothelial Cells\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "layout = {}\n",
    "    \n",
    "layout[\"height\"] = 800\n",
    "layout[\"width\"] = 1100\n",
    "layout[\"plot_bgcolor\"] = \"rgba(255, 255, 255, 0)\"\n",
    "layout[\"paper_bgcolor\"] = \"rgba(255, 255, 255, 0)\"\n",
    "layout[\"yaxis\"] = {\n",
    "    \"showgrid\": False,\n",
    "    \"zeroline\": False,\n",
    "    \"showticklabels\": False\n",
    "}\n",
    "layout[\"xaxis\"] = {\n",
    "    \"showgrid\": False,\n",
    "    \"zeroline\": False,\n",
    "    \"showticklabels\": False\n",
    "}\n",
    "layout[\"showlegend\"] = False\n",
    "\n",
    "layout = graph_objects.Layout(layout)\n",
    "\n",
    "figure = graph_objects.Figure(\n",
    "    data=[other_cells_trace, endothelial_trace],\n",
    "    layout=layout\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figure.write_image(os.path.join(\"out\", \"endothelial_cell_tsne.png\"), scale=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_sum = 10000\n",
    "adata_norm = adata.copy()\n",
    "X_norm_new = scanpy.pp.normalize_total(adata_norm, target_sum=target_sum, inplace=False)[\"X\"]\n",
    "adata_norm.X = X_norm_new\n",
    "scanpy.pp.log1p(adata_norm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scanpy.tl.rank_genes_groups(adata_norm, \"Cell Type\", method=\"wilcoxon\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "de_results = scanpy.get.rank_genes_groups_df(adata_norm, group=\"Endothelial Cells\")\n",
    "\n",
    "de_results[\"Gene Name\"] = adata.var.loc[de_results[\"names\"].values][\"Gene Name\"].values.astype(str)\n",
    "\n",
    "total_transcript_counts = numpy.array(adata.X.sum(axis=1)).flatten()\n",
    "adata_endothelial = adata[adata.obs[\"Cell Type\"] == \"Endothelial Cells\"]\n",
    "total_transcript_counts_endothelial = total_transcript_counts[adata.obs[\"Cell Type\"] == \"Endothelial Cells\"]\n",
    "\n",
    "# endothelial_expression = numpy.array(adata_endothelial[:, de_results[\"names\"].values.astype(str)].X.sum(axis=0)).flatten()\n",
    "# mean_expressions = endothelial_expression / total_transcript_counts_endothelial.sum()\n",
    "\n",
    "mean_expressions = numpy.divide(\n",
    "    numpy.array(adata_endothelial[:, de_results[\"names\"].values.astype(str)].X.todense()),\n",
    "    total_transcript_counts_endothelial.reshape((-1, 1)),\n",
    ").mean(axis=0)\n",
    "\n",
    "de_results[\"Mean Expression\"] = mean_expressions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def get_names_of_genes_with_specific_features(feature):\n",
    "    \"\"\"\"Retrieve names of genes that have a certain feature, such as \"membrane\" localized.\n",
    "    \n",
    "    Parameters:\n",
    "        feature (string): A single word in string format that defines the feature of interest (e.g. \"membrane\").\n",
    "        \n",
    "    Output:\n",
    "        genes_of_interest (dictionary): A dictionary listing all the names of genes that possess the feature of interest.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Define constants\n",
    "    BASE = 'http://www.uniprot.org'\n",
    "    KB_ENDPOINT = '/uniprot/'\n",
    "    TOOL_ENDPOINT = '/uploadlists/'\n",
    "\n",
    "    # Define query\n",
    "    payload = {'query': 'keyword:' + feature +' AND taxonomy:mus AND reviewed:yes',\n",
    "               'format': 'tab',\n",
    "               'columns': 'genes'}\n",
    "\n",
    "    # Pull requests\n",
    "    result = requests.get(BASE + KB_ENDPOINT, params=payload)\n",
    "\n",
    "    if result.ok:\n",
    "        print('Number of proteins with the keyword %s: ' % feature + str(len(result.text)))\n",
    "    else:\n",
    "        print('Something went wrong ', result.status_code)\n",
    "\n",
    "    # Create dictionary\n",
    "    genes_of_interest =  set()\n",
    "\n",
    "    # Iterate through words from UniProt output\n",
    "    for idx,row in enumerate(result.text.split()):\n",
    "\n",
    "        # First two elements are \"Gene names\"\n",
    "        if idx<3:\n",
    "            continue\n",
    "\n",
    "        # Add gene names to dictionary\n",
    "        genes_of_interest.add(row)\n",
    "        \n",
    "    return genes_of_interest\n",
    "\n",
    "membrane_genes = get_names_of_genes_with_specific_features(\"\\\"Cell membrane\\\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENE_GROUPS = {\n",
    "    \"Screen 1\": [\n",
    "        \"Ly6c1\",\n",
    "        \"Cldn5\",\n",
    "        \"Flt1\",\n",
    "        \"Ly6a\",\n",
    "        \"Slco1a4\",\n",
    "        \"Abcb1a\",\n",
    "        \"Bsg\",\n",
    "        \"Slc2a1\",\n",
    "        \"Esam\",\n",
    "        \"Adgrl4\",\n",
    "        \"Clec2d\",\n",
    "        \"Ifitm3\",\n",
    "        \"Adgrf5\",\n",
    "        \"Abcg2\",\n",
    "        \"Eng\",\n",
    "        \"Cdh5\",\n",
    "        \"Acvrl1\",\n",
    "        \"Pecam1\",\n",
    "        \"Slco1c1\",\n",
    "        \"Slc6a6\",\n",
    "        \"Fcgrt\",\n",
    "        \"Ocln\",\n",
    "        \"Kdr\",\n",
    "        \"Tek\",\n",
    "        \"Gpm6b\",\n",
    "        \"Gpm6a\",\n",
    "        \"Slc22a8\",\n",
    "        \"Podxl\",\n",
    "        \"Igf1r\",\n",
    "        \"Slc7a1\",\n",
    "        \"Prom1\",\n",
    "        \"Clic4\",\n",
    "        \"Car4\",\n",
    "        \"Clic1\",\n",
    "        \"Kitl\",\n",
    "        \"Cd34\",\n",
    "        \"Ly6e\",\n",
    "        \"Serinc3\",\n",
    "        \"App\",\n",
    "        \"Itm2b\",\n",
    "        \"Paqr5\",\n",
    "        \"Ifitm2\",\n",
    "      \n",
    "    ],\n",
    "    \"Screen 2\": [\n",
    "        \"Ly6h\",\n",
    "        \"Lypd5\",\n",
    "#         \"Cd59a\",\n",
    "        \"Car9\",\n",
    "        \"Car12\",\n",
    "        \"Car14\",\n",
    "        \"Car15\",\n",
    "        \"Ly6c1\",\n",
    "        \"Car4\",\n",
    "        \"Igf1r\"\n",
    "\n",
    "    ],\n",
    "    \"Hits\": [\n",
    "        \"Ly6a\",\n",
    "        \"Ly6c1\",\n",
    "        \"Car4\",\n",
    "        \"Igf1r\"\n",
    "    ]\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GROUP_COLORS = {\n",
    "    \"Screen 1\": \"red\",\n",
    "    \"Screen 2\": \"red\",\n",
    "    \"Hits\": \"red\",\n",
    "    \"Default\": \"black\"\n",
    "}\n",
    "\n",
    "GROUP_SIZES = {\n",
    "    \"Screen 1\": 10,\n",
    "    \"Screen 2\": 10,\n",
    "    \"Hits\": 10,\n",
    "    \"Default\": 5\n",
    "}\n",
    "\n",
    "GROUP_SHAPES = {\n",
    "    \"Screen 1\": \"circle\",\n",
    "    \"Screen 2\": \"circle\",\n",
    "    \"Hits\": \"cross\",\n",
    "    \"Default\": \"circle\"\n",
    "}\n",
    "\n",
    "GROUP_OPACITIES = {\n",
    "    \"Screen 1\": 0.25,\n",
    "    \"Screen 2\": 1,\n",
    "    \"Hits\": 1,\n",
    "    \"Default\": 0.25\n",
    "}\n",
    "\n",
    "gene_colors = {}\n",
    "\n",
    "for group, genes in GENE_GROUPS.items():\n",
    "    \n",
    "    for gene in genes:\n",
    "        \n",
    "        gene_colors[gene] = GROUP_COLORS[group]\n",
    "        \n",
    "gene_sizes = {}\n",
    "\n",
    "for group, genes in GENE_GROUPS.items():\n",
    "    \n",
    "    for gene in genes:\n",
    "        \n",
    "        gene_sizes[gene] = GROUP_SIZES[group]\n",
    "        \n",
    "gene_shapes = {}\n",
    "\n",
    "for group, genes in GENE_GROUPS.items():\n",
    "    \n",
    "    for gene in genes:\n",
    "        \n",
    "        gene_shapes[gene] = GROUP_SHAPES[group]\n",
    "        \n",
    "gene_opacities = {}\n",
    "\n",
    "for group, genes in GENE_GROUPS.items():\n",
    "    \n",
    "    for gene in genes:\n",
    "        \n",
    "        gene_opacities[gene] = GROUP_OPACITIES[group]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CUSTOM_MEMBRANE_GENES = [\"Cd34\"]\n",
    "\n",
    "CYTOPLASMIC_MEMBRANE_GENES = [\n",
    "    \"Ctnnb1\",\n",
    "    \"Ndrg1\",\n",
    "    \"Arl4a\",\n",
    "    \"Gng11\",\n",
    "    \"Vim\",\n",
    "    \"Sptbn1\",\n",
    "    \"Nostrin\",\n",
    "    \"Gng5\",\n",
    "    \"Gnai2\",\n",
    "    \"Rras\",\n",
    "    \"Rhoa\",\n",
    "    \"Rhoc\",\n",
    "    \"Lims2\",\n",
    "    \"Gnas\",\n",
    "    \"Eef1a1\",\n",
    "    \"Hsp90ab1\",\n",
    "    \"Slc9a3r2\",\n",
    "    \"Rpsa\",\n",
    "    \"Glul\",\n",
    "    \"Rhob\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_mask = \\\n",
    "    (\n",
    "        de_results[\"Gene Name\"].isin(membrane_genes) |\n",
    "        de_results[\"Gene Name\"].isin(CUSTOM_MEMBRANE_GENES)\n",
    "    ) & \\\n",
    "    (\n",
    "        ~de_results[\"Gene Name\"].isin(CYTOPLASMIC_MEMBRANE_GENES)\n",
    "    ) & \\\n",
    "    (de_results[\"scores\"] > 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "de_results_filtered = de_results[gene_mask].copy()\n",
    "\n",
    "color = [gene_colors[gene] if gene in gene_colors else GROUP_COLORS[\"Default\"] for gene in de_results_filtered[\"Gene Name\"]]\n",
    "sizes = [gene_sizes[gene] if gene in gene_sizes else GROUP_SIZES[\"Default\"] for gene in de_results_filtered[\"Gene Name\"]]\n",
    "shapes = [gene_shapes[gene] if gene in gene_shapes else GROUP_SHAPES[\"Default\"] for gene in de_results_filtered[\"Gene Name\"]]\n",
    "opacities = [gene_opacities[gene] if gene in gene_shapes else GROUP_OPACITIES[\"Default\"] for gene in de_results_filtered[\"Gene Name\"]]\n",
    "\n",
    "traces = []\n",
    "\n",
    "scatter_trace = graph_objects.Scatter(\n",
    "    y=de_results_filtered[\"Mean Expression\"],\n",
    "    x=de_results_filtered[\"scores\"],\n",
    "    mode=\"markers\",\n",
    "    text=de_results_filtered[\"Gene Name\"],\n",
    "    marker={\n",
    "        \"size\": sizes,\n",
    "        \"color\": color,\n",
    "        \"symbol\": shapes,\n",
    "        \"opacity\": opacities,\n",
    "        \"line\": {\n",
    "            \"color\": \"rgba(255, 255, 255, 0)\",\n",
    "            \"width\": 0\n",
    "        }\n",
    "    }\n",
    ")\n",
    "\n",
    "traces.append(scatter_trace)\n",
    "\n",
    "layout = {}\n",
    "\n",
    "layout[\"height\"] = 500\n",
    "layout[\"width\"] = 800\n",
    "layout[\"plot_bgcolor\"] = \"rgba(255, 255, 255, 0)\"\n",
    "layout[\"paper_bgcolor\"] = \"rgba(255, 255, 255, 0)\"\n",
    "layout[\"yaxis\"] = {\n",
    "    \"title\": \"Mean Transcript Abundance\",\n",
    "    \"type\": \"log\"\n",
    "}\n",
    "layout[\"xaxis\"] = {\n",
    "    \"title\": \"Score\"\n",
    "}\n",
    "\n",
    "layout = graph_objects.Layout(layout)\n",
    "\n",
    "figure = graph_objects.Figure(\n",
    "    data=traces,\n",
    "    layout=layout\n",
    ")\n",
    "\n",
    "plotly.iplot(figure)\n",
    "\n",
    "figure.write_image(os.path.join(\"out\", \"endothelial_cell_mean_abundance_vs_scanpy_score_upregulated_only_ly6a_marked.svg\"))\n",
    "figure.write_html(os.path.join(\"out\", \"endothelial_cell_mean_abundance_vs_scanpy_score_upregulated_only_ly6a_marked.html\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MARKER_GENE_LIST = [\"Cldn5\", \"Slc2a1\", \"Pecam1\", \"Cdh5\", \"Tek\", \"Ly6a\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_norm.obs[\"Is Endothelial\"] = adata_norm.obs[\"Cell Type\"] == \"Endothelial Cells\"\n",
    "adata_norm.obs.loc[adata_norm.obs[\"Cell Type\"] == \"Endothelial Cells\", \"Is Endothelial\"] = \"Endothelial Cells\"\n",
    "adata_norm.obs.loc[adata_norm.obs[\"Cell Type\"] != \"Endothelial Cells\", \"Is Endothelial\"] = \"Not Endothelial Cells\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figure = make_subplots(\n",
    "    rows=len(MARKER_GENE_LIST),\n",
    "    cols=2,\n",
    "    shared_xaxes=True,\n",
    "    shared_yaxes=True,\n",
    "    vertical_spacing=0,\n",
    "    horizontal_spacing=0.0\n",
    ")\n",
    "\n",
    "for gene_index, gene in enumerate(MARKER_GENE_LIST):\n",
    "    \n",
    "    gene_id = adata_norm.var[adata_norm.var[\"Gene Name\"] == gene].index.values[0]\n",
    "    \n",
    "    gene_max = -numpy.infty\n",
    "\n",
    "    for cell_type_index, cell_type in enumerate(sorted(adata_norm.obs[\"Is Endothelial\"].unique())):\n",
    "        \n",
    "        gene_counts = numpy.array(adata_norm[adata_norm.obs[\"Is Endothelial\"] == cell_type, gene_id].X.todense()).flatten()[::25]\n",
    "\n",
    "        x_values = [cell_type] * len(gene_counts)\n",
    "        y_values = gene_counts\n",
    "        \n",
    "        gene_max = max(gene_max, numpy.max(gene_counts))\n",
    "    \n",
    "        violin_trace = graph_objects.Violin(\n",
    "            x=x_values,\n",
    "            y=y_values,\n",
    "            line={\n",
    "                \"width\": 0.75,\n",
    "                \"color\": \"red\" if cell_type == \"Endothelial Cells\" else \"blue\"\n",
    "            },\n",
    "            marker={\n",
    "                \"size\": 1,\n",
    "                \"color\": \"red\" if cell_type == \"Endothelial Cells\" else \"blue\"\n",
    "            },\n",
    "            points=False,\n",
    "            hoverinfo=\"none\"\n",
    "        )\n",
    "\n",
    "        figure.add_trace(violin_trace, row=gene_index + 1, col=cell_type_index + 1)\n",
    "    \n",
    "        figure.update_xaxes(\n",
    "            {\n",
    "                \"tickangle\": -90\n",
    "            },\n",
    "            row=gene_index + 1,\n",
    "            col=cell_type_index + 1\n",
    "        )\n",
    "    \n",
    "    gene_max = numpy.ceil(gene_max * 4.4) / 4\n",
    "    gene_max = numpy.round(gene_max, 2)\n",
    "    \n",
    "    for cell_type_index, cell_type in enumerate(adata_norm.obs[\"Is Endothelial\"].unique()):\n",
    "        \n",
    "        figure.update_yaxes(\n",
    "            {\n",
    "                \"visible\": True,\n",
    "                \"range\": [-0.2*gene_max, gene_max]\n",
    "            },\n",
    "            row=gene_index + 1,\n",
    "            col=cell_type_index + 1\n",
    "        )\n",
    "\n",
    "    figure.update_yaxes(\n",
    "        {\n",
    "            \"visible\": True,\n",
    "            \"side\": \"right\",\n",
    "            \"tickvals\": [0, gene_max - 0.25],\n",
    "            \"range\": [-0.2*gene_max, gene_max]\n",
    "        },\n",
    "        row=gene_index + 1,\n",
    "        col=2\n",
    "    )\n",
    "    \n",
    "layout = {\n",
    "    \"plot_bgcolor\": \"rgba(0, 0, 0, 0)\",\n",
    "    \"showlegend\": False,\n",
    "    \"violingroupgap\": 0,\n",
    "    \"annotations\": [\n",
    "        graph_objects.layout.Annotation(\n",
    "            x=-0.4,\n",
    "            y=1 - (gene_index/len(MARKER_GENE_LIST) + 0.75/len(MARKER_GENE_LIST)),\n",
    "            xref=\"x\",\n",
    "            yref=\"paper\",\n",
    "            xanchor=\"right\",\n",
    "            text=gene,\n",
    "            showarrow=False,\n",
    "            align=\"right\"\n",
    "        )\n",
    "        for gene_index, gene in enumerate(MARKER_GENE_LIST)\n",
    "    ],\n",
    "    \"height\": 1500,\n",
    "    \"width\": 600,\n",
    "    \"margin\": graph_objects.layout.Margin(\n",
    "        l=100,\n",
    "        r=25\n",
    "    )\n",
    "}\n",
    "\n",
    "figure.update_layout(layout)\n",
    "\n",
    "plotly.iplot(figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figure.write_image(os.path.join(\"out\", \"endothelial_cell_marker_violin_plot.svg\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
