{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from aavomics import database\n",
    "import pandas\n",
    "import anndata\n",
    "import os\n",
    "import numpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Which alignment to use. Set to None to use the first available\n",
    "ALIGNMENT_NAME = \"cellranger_5.0.1_gex_mm10_2020_A\"\n",
    "VIRUS_ALIGNMENT_NAME = \"cellranger_5.0.1_gex_mm10_2020_A_AAVomics\"\n",
    "\n",
    "DROPLET_TAXONOMY_NAME = \"CCN202105041\"\n",
    "CELL_TYPE_TAXONOMY_NAME = \"CCN202105060\"\n",
    "\n",
    "\n",
    "AMPLIFIED_CELL_SETS_FILTER = [\"20181127_TC1\", \"20190319_TC2\", \"20190111_BC1\", \"20190321_BC2\", \"20190711_TC4\", \"20190712_TC5\", \"20190713_TC6\", \"20190713_TC7\", \"20200720_BC4_1\", \"20200720_BC4_2\", \"20210719_BC5\", \"20210720_BC6\"]\n",
    "INJECTED_CELL_SETS_FILTER = AMPLIFIED_CELL_SETS_FILTER + [\"20200807_C1\", \"20210726_TC11\", \"20210728_C5\", \"20210728_TC12\", \"20200907_C3\", \"20201119_C4\", \"20200903_TC8\", \"20200904_TC9\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_dict = {}\n",
    "\n",
    "for cell_set in database.CELL_SETS:\n",
    "    \n",
    "    if cell_set.source_tissue.animal is None:\n",
    "        continue\n",
    "    \n",
    "    if cell_set.source_tissue.region != \"Cortex\":\n",
    "        continue\n",
    "    \n",
    "    num_transcriptome_reads = 0\n",
    "    num_virus_reads = 0\n",
    "    \n",
    "    for sequencing_library in cell_set.sequencing_libraries:\n",
    "        for read_set in sequencing_library.read_sets:\n",
    "            if sequencing_library.type == \"Transcriptome\":\n",
    "                num_transcriptome_reads += read_set.num_reads\n",
    "                \n",
    "    if cell_set.name in AMPLIFIED_CELL_SETS_FILTER:\n",
    "        for sequencing_library in cell_set.sequencing_libraries:\n",
    "                if sequencing_library.type == \"Virus Transcripts\":\n",
    "                    num_virus_reads += read_set.num_reads\n",
    "    \n",
    "    adata_path = cell_set.get_anndata_file_path(alignment_name=ALIGNMENT_NAME, transcript_type=\"transcriptome\")\n",
    "    \n",
    "    if not os.path.exists(adata_path):\n",
    "        continue\n",
    "    \n",
    "    adata = anndata.read_h5ad(adata_path, backed=\"r\")\n",
    "    called_cells_mask = adata.obs[\"Cell Called\"] == \"True\"\n",
    "    typed_cells_mask = ~(adata.obs[CELL_TYPE_TAXONOMY_NAME].isna())\n",
    "    print(called_cells_mask.sum(), typed_cells_mask.sum())\n",
    "    cellranger_called_cells_mask = adata.obs[\"Cell Ranger Called\"] == True\n",
    "    multiplets_mask = adata.obs[DROPLET_TAXONOMY_NAME] == \"Multiplets\"\n",
    "    UMIs_per_cell = numpy.median(numpy.array(adata[called_cells_mask].X.sum(axis=1)).flatten())\n",
    "    genes_per_cell = numpy.median((adata[called_cells_mask].X > 0).toarray().sum(axis=1))\n",
    "    \n",
    "    num_called_cells = called_cells_mask.sum()\n",
    "    \n",
    "    vector_cargos = set()\n",
    "    viruses = set()\n",
    "        \n",
    "    for injection in cell_set.source_tissue.animal.injections:\n",
    "        for vector in injection.vector_pool.vectors:\n",
    "            vector_cargos.add(\"%s %s\" % (vector.delivery_vehicle.name, vector.cargo.name))\n",
    "            viruses.add(vector.delivery_vehicle.name)\n",
    "    \n",
    "    if len(viruses) == 0:\n",
    "        for injection in cell_set.source_tissue.animal.injections:\n",
    "            vector_cargos.add(injection.vector_pool.name)\n",
    "            \n",
    "    \n",
    "    if cell_set.name in INJECTED_CELL_SETS_FILTER:\n",
    "        vector_cargos = \";\".join(vector_cargos)\n",
    "        age_at_injection = (cell_set.source_tissue.animal.injection_date - cell_set.source_tissue.animal.DOB).days\n",
    "        incubation_time = (cell_set.source_tissue.animal.extraction_date - cell_set.source_tissue.animal.injection_date).days\n",
    "    else:\n",
    "        vector_cargos = \"Cell Typing Only\"\n",
    "        age_at_injection = 0\n",
    "        incubation_time = 0\n",
    "        \n",
    "    if cell_set.name in AMPLIFIED_CELL_SETS_FILTER:\n",
    "        virus_adata_path = cell_set.get_anndata_file_path(alignment_name=VIRUS_ALIGNMENT_NAME, transcript_type=\"transcriptome\")\n",
    "        virus_adata = anndata.read_h5ad(virus_adata_path, backed=\"r\")\n",
    "        num_disambiguated_virus_barcode_UMIs = virus_adata[called_cells_mask].obs.loc[:, viruses].sum().sum()\n",
    "        num_virus_barcode_UMIs = virus_adata[called_cells_mask, \"AAV\"].X.sum()\n",
    "        virus_disambiguation_rate = num_disambiguated_virus_barcode_UMIs/num_virus_barcode_UMIs\n",
    "    else:\n",
    "        virus_disambiguation_rate = 0\n",
    "        \n",
    "    if cell_set.source_tissue.animal.DOB is None:\n",
    "        age_at_extraction = 0\n",
    "    else:\n",
    "        age_at_extraction = (cell_set.source_tissue.animal.extraction_date - cell_set.source_tissue.animal.DOB).days\n",
    "        \n",
    "    samples_dict[cell_set.name] = {\n",
    "        \"10X Version\": int(cell_set.source_tissue.dissociation_run.protocol_version),\n",
    "        \"Animal ID\": cell_set.source_tissue.animal.name,\n",
    "        \"Target # Cells\": int(cell_set.target_num_cells),\n",
    "        \"Recovered Cells\": num_called_cells,\n",
    "        \"Typed Cells\": typed_cells_mask.sum(),\n",
    "        \"Cell Ranger # Cells\": cellranger_called_cells_mask.sum(),\n",
    "        \"Predicted Multiplets\": multiplets_mask.sum(),\n",
    "        \"Transcriptome Sequencing Depth\": num_transcriptome_reads,\n",
    "        \"Transcriptome Reads/Cell\": num_transcriptome_reads/num_called_cells,\n",
    "        \"Median UMIs/Cell\": UMIs_per_cell,\n",
    "        \"Median Genes/Cell\": genes_per_cell,\n",
    "        \"Variants Recovered\": vector_cargos,\n",
    "        \"Virus Sequencing Depth\": num_virus_reads,\n",
    "        \"Virus Reads/Cell\": num_virus_reads/num_called_cells,\n",
    "        \"Age at Extraction (Days)\": age_at_extraction,\n",
    "        \"Virus Incubation time (Days)\": incubation_time,\n",
    "        \"Percent of Virus UMIs Determined\": virus_disambiguation_rate * 100\n",
    "    }\n",
    "    \n",
    "    display(pandas.DataFrame.from_dict(samples_dict, orient=\"index\"))\n",
    "    \n",
    "df = pandas.DataFrame.from_dict(samples_dict, orient=\"index\")\n",
    "df.to_csv(os.path.join(database.DATA_PATH, \"aavomics_sample_metadata.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
