{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from aavomics import database\n",
    "import os\n",
    "import pandas\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_sets_dict = {}\n",
    "\n",
    "for read_set in database.READ_SETS:\n",
    "    \n",
    "    read_set_name = \"AAVomics_\" + read_set.name\n",
    "    read_sets_dict[read_set_name] = {}\n",
    "    \n",
    "    animal = None\n",
    "    \n",
    "    sequencing_library = read_set.sequencing_libraries[0]\n",
    "    \n",
    "    if sequencing_library.type != \"Transcriptome\":\n",
    "        continue\n",
    "    \n",
    "    tissue = sequencing_library.cell_set.source_tissue.region\n",
    "    animal = sequencing_library.cell_set.source_tissue.animal\n",
    "    \n",
    "    read_sets_dict[read_set_name][\"organism\"] = \"Mus musculus\"\n",
    "    read_sets_dict[read_set_name][\"tissue\"] = tissue\n",
    "    read_sets_dict[read_set_name][\"sex\"] = \"male\"\n",
    "    read_sets_dict[read_set_name][\"birth_date\"] = animal.DOB\n",
    "    read_sets_dict[read_set_name][\"collection_date\"] = animal.extraction_date\n",
    "    read_sets_dict[read_set_name][\"animal_id\"] = animal.name\n",
    "    if animal.DOB is not None:\n",
    "        read_sets_dict[read_set_name][\"age\"] = (animal.extraction_date - animal.DOB).days\n",
    "    else:\n",
    "        read_sets_dict[read_set_name][\"dev_stage\"] = \"adult\"\n",
    "    \n",
    "df = pandas.DataFrame.from_dict(read_sets_dict, orient=\"index\")\n",
    "df.to_csv(os.path.join(database.DATA_PATH, \"aavomics_SRA_sample_metadata.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_sets_dict = {}\n",
    "\n",
    "for read_set in database.READ_SETS:\n",
    "    \n",
    "    read_set_name = \"AAVomics_\" + read_set.name\n",
    "    read_sets_dict[read_set_name] = {}\n",
    "    \n",
    "    animal = None\n",
    "    \n",
    "    sequencing_library = read_set.sequencing_libraries[0]\n",
    "    \n",
    "    if sequencing_library.type == \"Transcriptome\":\n",
    "        read_sets_dict[read_set_name][\"library_selection\"] = \"RANDOM\"\n",
    "        files = os.listdir(os.path.join(database.DATA_PATH, \"cell_sets\", read_set.sequencing_libraries[0].cell_set.name, \"transcriptome\", \"reads\"))\n",
    "    else:\n",
    "        read_sets_dict[read_set_name][\"library_selection\"] = \"PCR\"\n",
    "        files = os.listdir(os.path.join(database.DATA_PATH, \"cell_sets\", read_set.sequencing_libraries[0].cell_set.name, \"virus\", \"reads\"))\n",
    "    \n",
    "    filename_index = 0\n",
    "    \n",
    "    for file in files:\n",
    "        \n",
    "        prefix = \"_\".join(file.split(\"_\")[0:-4])\n",
    "        \n",
    "        if prefix == read_set.name:\n",
    "            if filename_index == 0:\n",
    "                read_sets_dict[read_set_name][\"filename\"] = file\n",
    "                filename_index += 1\n",
    "            else:\n",
    "                read_sets_dict[read_set_name][\"filename%i\" % (filename_index + 1)] = file\n",
    "                filename_index += 1\n",
    "        \n",
    "    read_sets_dict[read_set_name][\"sample_name\"] = \"AAVomics_\" + \"_\".join(read_set.sequencing_libraries[0].cell_set.name.split(\"_\")[1:])\n",
    "    read_sets_dict[read_set_name][\"type\"] = read_set.sequencing_run.type\n",
    "    \n",
    "df = pandas.DataFrame.from_dict(read_sets_dict, orient=\"index\")\n",
    "df.to_csv(os.path.join(database.DATA_PATH, \"aavomics_SRA_file_metadata.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "UPLOAD_FOLDER = os.path.join(database.DATA_PATH, \"SRA_upload\")\n",
    "\n",
    "if not os.path.exists(UPLOAD_FOLDER):\n",
    "    os.makedirs(UPLOAD_FOLDER)\n",
    "\n",
    "for read_set in database.READ_SETS:\n",
    "    \n",
    "    read_set_name = \"AAVomics_\" + read_set.name\n",
    "    read_sets_dict[read_set_name] = {}\n",
    "    \n",
    "    sequencing_library = read_set.sequencing_libraries[0]\n",
    "    \n",
    "    if sequencing_library.type == \"Transcriptome\":\n",
    "        read_sets_dict[read_set_name][\"library_selection\"] = \"RANDOM\"\n",
    "        reads_dir = os.path.join(database.DATA_PATH, \"cell_sets\", read_set.sequencing_libraries[0].cell_set.name, \"transcriptome\", \"reads\")\n",
    "\n",
    "    else:\n",
    "        read_sets_dict[read_set_name][\"library_selection\"] = \"PCR\"\n",
    "        reads_dir = os.path.join(database.DATA_PATH, \"cell_sets\", read_set.sequencing_libraries[0].cell_set.name, \"virus\", \"reads\")\n",
    "    \n",
    "    files = os.listdir(reads_dir)\n",
    "    filename_index = 0\n",
    "    \n",
    "    files_to_upload = []\n",
    "    \n",
    "    for file in files:\n",
    "        \n",
    "        prefix = \"_\".join(file.split(\"_\")[0:-4])\n",
    "        \n",
    "        if prefix == read_set.name:\n",
    "            shutil.copy(os.path.join(reads_dir, file), os.path.join(UPLOAD_FOLDER, file))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
