{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy \n",
    "\n",
    "from aavomics import database\n",
    "import scanpy\n",
    "import anndata\n",
    "import pandas\n",
    "\n",
    "from aavomics import aavomics\n",
    "\n",
    "from plotly import offline as plotly\n",
    "from plotly import graph_objects\n",
    "from plotly.subplots import make_subplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALIGNMENT_NAME = \"cellranger_5.0.1_gex_mm10_2020_A_AAVomics\"\n",
    "TAXONOMY_ALIGNMENT_NAME = \"cellranger_5.0.1_gex_mm10_2020_A\"\n",
    "\n",
    "CELL_TYPE_TAXONOMY_NAME = \"CCN202105060\"\n",
    "TAXONOMY_NAMES = [CELL_TYPE_TAXONOMY_NAME] + [\"CCN202105041\", \"CCN202105050\", \"CCN202105051\", \"CCN202105070\"]\n",
    "CELL_TYPES_TO_FILTER = [\"Debris\", \"Multiplets\", \"Unknown\", \"\", \"nan\", numpy.nan, \"NaN\"]\n",
    "TAXONOMY_METADATA_PREFIXES = \"p\", \"c\", \"X\", \"Y\"\n",
    "\n",
    "ANNDATA_FILE_NAME = \"aavomics_mouse_cortex_2021.h5ad\"\n",
    "METADATA_FILE_NAME = \"aavomics_cell_metadata.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adatas = []\n",
    "cell_set_names = []\n",
    "\n",
    "for cell_set_index, cell_set in enumerate(database.CELL_SETS):\n",
    "    \n",
    "    print(\"Adding %s\" % cell_set.name)\n",
    "    \n",
    "    tissue_sample = cell_set.source_tissue\n",
    "    animal = tissue_sample.animal\n",
    "    \n",
    "    adata = anndata.read_h5ad(cell_set.get_anndata_file_path(alignment_name=ALIGNMENT_NAME))\n",
    "    \n",
    "    taxonomy_adata = anndata.read_h5ad(cell_set.get_anndata_file_path(alignment_name=TAXONOMY_ALIGNMENT_NAME))\n",
    "    \n",
    "    cell_mask = numpy.zeros((taxonomy_adata.shape[0], ), dtype=numpy.bool)\n",
    "\n",
    "    for taxonomy_name in TAXONOMY_NAMES:\n",
    "\n",
    "        taxonomy_mask = numpy.ones((taxonomy_adata.shape[0], ), dtype=numpy.bool)\n",
    "\n",
    "        taxonomy_mask = taxonomy_mask & (~taxonomy_adata.obs[taxonomy_name].isna())    \n",
    "        for cell_type_to_filter in CELL_TYPES_TO_FILTER:\n",
    "            taxonomy_mask = taxonomy_mask & (taxonomy_adata.obs[taxonomy_name] != cell_type_to_filter)\n",
    "\n",
    "        cell_mask = cell_mask | taxonomy_mask\n",
    "        \n",
    "    fresh_adata = taxonomy_adata[cell_mask].copy()\n",
    "\n",
    "    for column in fresh_adata.obs.columns:\n",
    "        fresh_adata.obs.drop(column, axis=1, inplace=True)\n",
    "    \n",
    "    injections = animal.injections\n",
    "\n",
    "    read_sets = set()\n",
    "\n",
    "    for sequencing_library in cell_set.sequencing_libraries:\n",
    "        if sequencing_library.type == \"Virus Transcripts\":\n",
    "            read_sets.update(sequencing_library.read_sets)\n",
    "    \n",
    "    virus_and_vector_names = set()\n",
    "    \n",
    "    if len(read_sets) == 0:\n",
    "        disambiguate_viruses = False\n",
    "        print(\"No amplified reads to disambiguate. Only including overall transduction\")\n",
    "    else:\n",
    "        disambiguate_viruses = True\n",
    "    \n",
    "    virus_vector_names = {}\n",
    "\n",
    "    for injection in injections:\n",
    "\n",
    "        for vector in injection.vector_pool.vectors:\n",
    "\n",
    "            virus = vector.delivery_vehicle\n",
    "\n",
    "            if virus.name not in virus_vector_names:\n",
    "                virus_vector_names[virus.name] = set([vector.name])\n",
    "            else:\n",
    "                virus_vector_names[virus.name].add(vector.name)\n",
    "\n",
    "    for virus_name, vector_names in virus_vector_names.items():\n",
    "\n",
    "        virus_and_vector_names.add(virus_name)\n",
    "\n",
    "        if len(vector_names) > 1:\n",
    "            virus_and_vector_names.update(vector_names)\n",
    "    \n",
    "    fresh_adata.obs.loc[fresh_adata.obs.index.values, \"Cell Type\"] = taxonomy_adata.obs.loc[fresh_adata.obs.index.values][CELL_TYPE_TAXONOMY_NAME]\n",
    "    \n",
    "    if len(virus_and_vector_names) > 0:\n",
    "        fresh_adata.obs.loc[fresh_adata.obs.index.values, \"AAV\"] = adata[fresh_adata.obs.index.values, \"AAV\"].X.todense()\n",
    "        \n",
    "    cell_set_names.append(cell_set.name)\n",
    "    \n",
    "    if disambiguate_viruses:\n",
    "        for column_name in virus_and_vector_names:\n",
    "            fresh_adata.obs.loc[fresh_adata.obs.index.values, column_name] = adata.obs.loc[fresh_adata.obs.index.values][column_name]\n",
    "        \n",
    "    for taxonomy_name in TAXONOMY_NAMES:\n",
    "        \n",
    "        if taxonomy_name != CELL_TYPE_TAXONOMY_NAME:\n",
    "            fresh_adata.obs.loc[fresh_adata.obs.index.values, taxonomy_name] = taxonomy_adata.obs.loc[fresh_adata.obs.index.values][taxonomy_name]\n",
    "        \n",
    "        for prefix in TAXONOMY_METADATA_PREFIXES:\n",
    "            \n",
    "            column_name = \"%s_%s\" % (prefix, taxonomy_name)\n",
    "            \n",
    "            if column_name in taxonomy_adata.obs.columns:\n",
    "                fresh_adata.obs.loc[fresh_adata.obs.index.values, column_name] = taxonomy_adata.obs.loc[fresh_adata.obs.index.values][column_name]\n",
    "    \n",
    "    fresh_adata.obs.index = [\"%s-%i\" % (x.split(\"-\")[0], cell_set_index + 1) for x in fresh_adata.obs.index.values]\n",
    "    \n",
    "    # Replace all NaN-like with numpy NaN\n",
    "    fresh_adata.obs.replace(\"nan\", numpy.nan, inplace=True)\n",
    "    fresh_adata.obs.replace(\"\", numpy.nan, inplace=True)\n",
    "    fresh_adata.obs.replace(\"Unknown\", numpy.nan, inplace=True)\n",
    "    display(fresh_adata.obs)\n",
    "    \n",
    "    adatas.append(fresh_adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_adata = anndata.concat(adatas, label=\"Cell Set\", keys=cell_set_names, join=\"outer\")\n",
    "combined_adata.var[\"Gene Name\"] = adata.var[\"Gene Name\"].loc[combined_adata.var.index.values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_adata.obs.replace(\"nan\", numpy.nan, inplace=True)\n",
    "combined_adata.obs.replace(\"\", numpy.nan, inplace=True)\n",
    "combined_adata.obs.replace(\"Unknown\", numpy.nan, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_adata.write_h5ad(os.path.join(database.DATA_PATH, ANNDATA_FILE_NAME))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_adata.obs.to_csv(os.path.join(database.DATA_PATH, METADATA_FILE_NAME))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
