{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "from aavomics import database\n",
    "from aavomics import aavomics\n",
    "import scrublet\n",
    "\n",
    "import scvi\n",
    "import scanpy\n",
    "import numpy\n",
    "import pandas\n",
    "import statsmodels\n",
    "from statsmodels.stats import multitest\n",
    "\n",
    "from skimage.filters import threshold_otsu\n",
    "from statsmodels.stats.proportion import proportions_ztest\n",
    "\n",
    "import anndata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SEED = 1042"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Which alignment to use. Set to None to use the first available\n",
    "ALIGNMENT_NAME = \"cellranger_5.0.1_gex_mm10_2020_A\"\n",
    "\n",
    "TAXONOMY_NAME = \"CCN202105041\"\n",
    "SUBTYPE_TAXONOMY_NAME = \"CCN202105051\"\n",
    "\n",
    "CLUSTER_OBS_NAME = \"leiden_scVI\"\n",
    "\n",
    "TRANSFORM_TO_PLOT = \"X_tsne\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adatas = []\n",
    "\n",
    "for cell_set in database.CELL_SETS:\n",
    "    \n",
    "    anndata_file_path = cell_set.get_anndata_file_path(alignment_name=ALIGNMENT_NAME)\n",
    "    \n",
    "    if not os.path.exists(anndata_file_path):\n",
    "        print(\"Skipping %s, anndata doesn't exist\" % cell_set.name)\n",
    "        continue\n",
    "    \n",
    "    adata = anndata.read(anndata_file_path)\n",
    "    \n",
    "    adata = adata[(adata.obs[TAXONOMY_NAME] == \"Non-Neurons\") & (adata.obs[\"Cell Called\"] == \"True\")].copy()\n",
    "    adata.obs[\"Cell Set\"] = cell_set.name\n",
    "    \n",
    "    print(cell_set.name, adata.shape[0])\n",
    "    \n",
    "    adatas.append(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adatas[0].concatenate(adatas[1:])\n",
    "\n",
    "gene_mask = (adata.X.max(axis=0) > 0).toarray().flatten()\n",
    "adata = adata[:, gene_mask].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scvi.data.setup_anndata(adata, batch_key=\"Cell Set\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vae = scvi.model.SCVI(\n",
    "    adata,\n",
    "    n_latent=20,\n",
    "    n_layers=2,\n",
    "    n_hidden=256\n",
    ")\n",
    "\n",
    "vae.train(\n",
    "    seed=SEED,\n",
    "    frequency=1,\n",
    "    n_epochs_kl_warmup=None,\n",
    "    n_iter_kl_warmup=128*5000/400, # Based on documentation at https://www.scvi-tools.org/en/stable/api/reference/scvi.core.trainers.UnsupervisedTrainer.html\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_expression = vae.get_normalized_expression(vae.adata)\n",
    "adata.obsm[\"X_scVI\"] = vae.get_latent_representation(vae.adata)\n",
    "scanpy.pp.neighbors(vae.adata, use_rep=\"X_scVI\", random_state=SEED)\n",
    "scanpy.tl.tsne(vae.adata, use_rep=\"X_scVI\", n_jobs=16, random_state=SEED)\n",
    "scanpy.tl.leiden(vae.adata, key_added=CLUSTER_OBS_NAME, random_state=SEED, resolution=2) # Resolution 2 to distinguish between doublet clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aavomics.plot_clusters(adata.obsm[TRANSFORM_TO_PLOT], adata.obs[CLUSTER_OBS_NAME], filename=os.path.join(\"out\", \"clusters_non_neuronal.html\"))\n",
    "aavomics.plot_clusters(adata.obsm[TRANSFORM_TO_PLOT], adata.obs[\"Cell Set\"], filename=os.path.join(\"out\", \"samples_non_neuronal.html\"))\n",
    "total_transcript_counts = numpy.array(adata.X.sum(axis=1)).flatten()\n",
    "aavomics.plot_gene_expression(adata.obsm[TRANSFORM_TO_PLOT], total_transcript_counts, filename=os.path.join(\"out\", \"total_transcript_counts_non_neuronal.html\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import numpy as np\n",
    "from scvi import _CONSTANTS\n",
    "from scvi.core.distributions import NegativeBinomial, ZeroInflatedNegativeBinomial\n",
    "from typing import Dict, Iterable, Optional, Sequence, Union\n",
    "from anndata import AnnData\n",
    "\n",
    "from scvi.model._utils import (\n",
    "    _get_batch_code_from_category,\n",
    "    _get_var_names_from_setup_anndata,\n",
    "    scrna_raw_counts_properties,\n",
    ")\n",
    "\n",
    "@torch.no_grad()\n",
    "def posterior_predictive_sample(\n",
    "    self,\n",
    "    adata: Optional[AnnData] = None,\n",
    "    indices: Optional[Sequence[int]] = None,\n",
    "    n_samples: int = 1,\n",
    "    gene_list: Optional[Sequence[str]] = None,\n",
    "    batch_size: Optional[int] = None,\n",
    "    transform_batch: Optional[int] = None\n",
    ") -> np.ndarray:\n",
    "    r\"\"\"\n",
    "    Generate observation samples from the posterior predictive distribution.\n",
    "\n",
    "    The posterior predictive distribution is written as :math:`p(\\hat{x} \\mid x)`.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata\n",
    "        AnnData object with equivalent structure to initial AnnData. If `None`, defaults to the\n",
    "        AnnData object used to initialize the model.\n",
    "    indices\n",
    "        Indices of cells in adata to use. If `None`, all cells are used.\n",
    "    n_samples\n",
    "        Number of samples for each cell.\n",
    "    gene_list\n",
    "        Names of genes of interest.\n",
    "    batch_size\n",
    "        Minibatch size for data loading into model. Defaults to `scvi.settings.batch_size`.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    x_new : :py:class:`torch.Tensor`\n",
    "        tensor with shape (n_cells, n_genes, n_samples)\n",
    "    \"\"\"\n",
    "    if self.model.gene_likelihood not in [\"zinb\", \"nb\", \"poisson\"]:\n",
    "        raise ValueError(\"Invalid gene_likelihood.\")\n",
    "\n",
    "    adata = self._validate_anndata(adata)\n",
    "    scdl = self._make_scvi_dl(adata=adata, indices=indices, batch_size=batch_size)\n",
    "\n",
    "    if indices is None:\n",
    "        indices = np.arange(adata.n_obs)\n",
    "\n",
    "    if gene_list is None:\n",
    "        gene_mask = slice(None)\n",
    "    else:\n",
    "        all_genes = _get_var_names_from_setup_anndata(adata)\n",
    "        gene_mask = [True if gene in gene_list else False for gene in all_genes]\n",
    "\n",
    "    x_new = []\n",
    "    for tensors in scdl:\n",
    "        x = tensors[_CONSTANTS.X_KEY]\n",
    "        batch_idx = tensors[_CONSTANTS.BATCH_KEY]\n",
    "        labels = tensors[_CONSTANTS.LABELS_KEY]\n",
    "        outputs = self.model.inference(\n",
    "            x, batch_index=batch_idx, y=labels, n_samples=n_samples, transform_batch=transform_batch\n",
    "        )\n",
    "        px_r = outputs[\"px_r\"]\n",
    "        px_rate = outputs[\"px_rate\"]\n",
    "        px_dropout = outputs[\"px_dropout\"]\n",
    "\n",
    "        if self.model.gene_likelihood == \"poisson\":\n",
    "            l_train = px_rate\n",
    "            l_train = torch.clamp(l_train, max=1e8)\n",
    "            dist = torch.distributions.Poisson(\n",
    "                l_train\n",
    "            )  # Shape : (n_samples, n_cells_batch, n_genes)\n",
    "        elif self.model.gene_likelihood == \"nb\":\n",
    "            dist = NegativeBinomial(mu=px_rate, theta=px_r)\n",
    "        elif self.model.gene_likelihood == \"zinb\":\n",
    "            dist = ZeroInflatedNegativeBinomial(\n",
    "                mu=px_rate, theta=px_r, zi_logits=px_dropout\n",
    "            )\n",
    "        else:\n",
    "            raise ValueError(\n",
    "                \"{} reconstruction error not handled right now\".format(\n",
    "                    self.model.gene_likelihood\n",
    "                )\n",
    "            )\n",
    "        if n_samples > 1:\n",
    "            exprs = dist.sample().permute(\n",
    "                [1, 2, 0]\n",
    "            )  # Shape : (n_cells_batch, n_genes, n_samples)\n",
    "        else:\n",
    "            exprs = dist.sample()\n",
    "\n",
    "        if gene_list is not None:\n",
    "            exprs = exprs[:, gene_mask, ...]\n",
    "\n",
    "        x_new.append(exprs.cpu())\n",
    "    x_new = torch.cat(x_new)  # Shape (n_cells, n_genes, n_samples)\n",
    "\n",
    "    return x_new.numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters = sorted(adata.obs[CLUSTER_OBS_NAME].unique(), key=lambda x: int(x))\n",
    "\n",
    "cluster_marker_p_non_zeros = {}\n",
    "\n",
    "NUM_SAMPLES = 5000\n",
    "\n",
    "for cluster in clusters:\n",
    "    \n",
    "    print(cluster)\n",
    "\n",
    "    adata_copy = adata[adata.obs[CLUSTER_OBS_NAME] == cluster].copy()\n",
    "    \n",
    "    p_non_zeros = []\n",
    "\n",
    "    for batch_id in adata_copy.obs[\"_scvi_batch\"].unique():\n",
    "        \n",
    "        random_sample = adata_copy[numpy.random.choice(list(range(0, adata_copy.shape[0])), replace=True, size=NUM_SAMPLES)].copy()\n",
    "\n",
    "        samples = posterior_predictive_sample(vae, random_sample, n_samples=1, transform_batch=batch_id)\n",
    "        \n",
    "        non_zero_counts = (samples != 0).sum(axis=0)\n",
    "        \n",
    "        p_non_zeros.append(non_zero_counts/NUM_SAMPLES)\n",
    "        \n",
    "    p_non_zeros = numpy.array(p_non_zeros).mean(axis=0)\n",
    "    \n",
    "    cluster_marker_p_non_zeros[cluster] = p_non_zeros\n",
    "    \n",
    "cluster_gene_non_zeros_df = pandas.DataFrame.from_dict(cluster_marker_p_non_zeros, orient=\"index\", columns=adata.var.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CELL_TYPE_MARKER_GENES = {\n",
    "    \"Vascular Cells\": {\n",
    "        \"Endothelial Cells\": {\n",
    "            \"Slc2a1\": 1,\n",
    "            \"Cldn5\": 1\n",
    "        },\n",
    "        \"Pericytes\": {\n",
    "            \"Pdgfrb\": 1,\n",
    "            \"Rgs5\": 1,\n",
    "            \"Abcc9\": 1\n",
    "        },\n",
    "        \"Red Blood Cells\": {\n",
    "            \"Hba-a1\": 1,\n",
    "            \"Hba-a2\": 1\n",
    "        },\n",
    "        \"Vascular SMCs\": {\n",
    "            \"Acta2\": 1,\n",
    "            \"Myh11\": 1,\n",
    "            \"Tagln\": 1\n",
    "        },\n",
    "        \"VLMCs\": {\n",
    "            \"Vtn\": 1,\n",
    "            \"Lum\": 1,\n",
    "            \"Col1a2\": 1,\n",
    "            \"Tbx18\": 1\n",
    "        }\n",
    "    },\n",
    "    \"Immune Cells\": {\n",
    "        \"Microglia\": {\n",
    "            \"Tmem119\": 1,\n",
    "            \"Cx3cr1\": 1,\n",
    "            \"P2ry12\": 1\n",
    "        },\n",
    "        \"Leukocytes\": {\n",
    "            \"Itgal\": 1,\n",
    "            \"Gzma\": 1\n",
    "        },\n",
    "        \"Perivascular Macrophages\": {\n",
    "            \"Mrc1\": 1\n",
    "        }\n",
    "    },\n",
    "    \"Oligodendrocytes\": {\n",
    "        \"OPCs\": {\n",
    "            \"Pdgfra\": 1,\n",
    "            \"Cspg4\": 1\n",
    "        },\n",
    "        \"Mature Oligodendrocytes\": {\n",
    "            \"Mag\": 1,\n",
    "            \"Mbp\": 1\n",
    "        },\n",
    "        \"Committed Oligodendrocytes\": {\n",
    "            \"Ptprz1\": 1,\n",
    "            \"Bmp4\": 1,\n",
    "            \"Nkx2-2\": 1,\n",
    "            \"Vcan\": 1\n",
    "        },\n",
    "        \"VLMCs\": {\n",
    "            \"Vtn\": 1,\n",
    "            \"Lum\": 1,\n",
    "            \"Col1a2\": 1,\n",
    "            \"Tbx18\": 1\n",
    "        }\n",
    "    },\n",
    "    \"Astrocytes\": {\n",
    "        \"Astrocytes\": {\n",
    "            \"Sox9\": 1\n",
    "        }\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_cell_type_label = {}\n",
    "\n",
    "for cluster in numpy.unique(cluster_gene_non_zeros_df.index):\n",
    "\n",
    "    possible_cell_types = set()\n",
    "\n",
    "    for cell_type in CELL_TYPE_MARKER_GENES:\n",
    "        \n",
    "        possible_cell_subtypes = set()\n",
    "        max_num_conditions_met = 0\n",
    "\n",
    "        for cell_subtype in CELL_TYPE_MARKER_GENES[cell_type]:\n",
    "\n",
    "            num_conditions_met = 0\n",
    "\n",
    "            for marker_gene in CELL_TYPE_MARKER_GENES[cell_type][cell_subtype]:\n",
    "\n",
    "                values = []\n",
    "\n",
    "                ensembl_id = adata.var.loc[adata.var['Gene Name']==marker_gene].index[0]\n",
    "                values = cluster_gene_non_zeros_df.loc[:, ensembl_id].values.flatten()\n",
    "                cluster_value = cluster_gene_non_zeros_df.loc[cluster, ensembl_id]\n",
    "\n",
    "                threshold = threshold_otsu(numpy.array(values))\n",
    "\n",
    "                if cluster_value > threshold and CELL_TYPE_MARKER_GENES[cell_type][cell_subtype][marker_gene] == 1:\n",
    "                    num_conditions_met += 1\n",
    "                    print(\"Cluster %s is expressing %s\" % (cluster, marker_gene))\n",
    "                elif cluster_value <= threshold and CELL_TYPE_MARKER_GENES[cell_type][cell_subtype][marker_gene] == 0:\n",
    "                    num_conditions_met += 1\n",
    "\n",
    "            if num_conditions_met == len(CELL_TYPE_MARKER_GENES[cell_type][cell_subtype]):\n",
    "                if num_conditions_met > max_num_conditions_met:\n",
    "                    max_num_conditions_met = num_conditions_met\n",
    "                    possible_cell_subtypes = set([cell_subtype])\n",
    "                elif num_conditions_met == max_num_conditions_met:\n",
    "                    possible_cell_subtypes.add(cell_subtype)\n",
    "                    \n",
    "        possible_cell_types.update(possible_cell_subtypes)\n",
    "\n",
    "    print(\"Cluster %s could be\" % cluster, possible_cell_types)\n",
    "\n",
    "    if len(possible_cell_types) > 1:\n",
    "        cluster_cell_type_label[cluster] = \"Multiplets\"\n",
    "    elif len(possible_cell_types) == 0:\n",
    "        cluster_cell_type_label[cluster] = \"Unknown\"\n",
    "    else:\n",
    "        cluster_cell_type_label[cluster] = list(possible_cell_types)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_cluster_cell_type_label = {}\n",
    "\n",
    "separate_clusters = set()\n",
    "\n",
    "cluster_marker_genes = {}\n",
    "\n",
    "for cluster, cell_type_label in cluster_cell_type_label.items():\n",
    "    \n",
    "    print(cluster, cell_type_label)\n",
    "    \n",
    "    if cell_type_label == \"Multiplets\" or cell_type_label == \"Unknown\":\n",
    "        continue\n",
    "    \n",
    "    other_clusters = []\n",
    "    \n",
    "    for other_cluster, other_cell_type_label in cluster_cell_type_label.items():\n",
    "        if other_cell_type_label == cluster_cell_type_label[cluster] and cluster != other_cluster:\n",
    "            other_clusters.append(other_cluster)\n",
    "                    \n",
    "    if len(other_clusters) == 0:\n",
    "        continue\n",
    "        \n",
    "    genes_above_threshold = cluster_marker_p_non_zeros[cluster] > 0.5\n",
    "    \n",
    "    other_cluster_num_non_zero = 0\n",
    "    num_cells_other_clusters = 0\n",
    "    \n",
    "    for other_cluster in other_clusters:\n",
    "        genes_above_threshold = genes_above_threshold & (cluster_marker_p_non_zeros[other_cluster] < cluster_marker_p_non_zeros[cluster]/10)\n",
    "        \n",
    "        num_cells_other_cluster = (adata.obs[CLUSTER_OBS_NAME] == other_cluster).sum()\n",
    "        num_cells_other_clusters += num_cells_other_cluster\n",
    "        other_cluster_num_non_zero += num_cells_other_cluster * cluster_marker_p_non_zeros[other_cluster]\n",
    "        \n",
    "    genes = adata.var.copy()\n",
    "    genes[\"p_non_zero\"] = cluster_marker_p_non_zeros[cluster]\n",
    "    genes[\"other_clusters_p_non_zero\"] = other_cluster_num_non_zero/num_cells_other_clusters\n",
    "    \n",
    "    \n",
    "    genes = genes[genes_above_threshold]\n",
    "    \n",
    "    num_cluster_cells = (adata.obs[CLUSTER_OBS_NAME] == cluster).sum()\n",
    "    \n",
    "    zs = []\n",
    "    \n",
    "    for gene in genes.iterrows():\n",
    "        z, p = proportions_ztest(\n",
    "            [gene[1][\"p_non_zero\"] * num_cluster_cells, gene[1][\"other_clusters_p_non_zero\"] * num_cells_other_clusters],\n",
    "            [num_cluster_cells, num_cells_other_clusters]\n",
    "        )\n",
    "        \n",
    "        zs.append(z)\n",
    "    \n",
    "    if len(genes) > 0:\n",
    "        separate_clusters.add(cluster)\n",
    "        genes[\"2proportionz\"] = zs\n",
    "        \n",
    "        genes = genes.sort_values(by=\"p_non_zero\", ascending=False)\n",
    "        \n",
    "        display(genes)\n",
    "        \n",
    "        top_gene = genes.iloc[0][\"Gene Name\"]\n",
    "        \n",
    "        new_cluster_cell_type_label[cluster] = \"%s+ %s\" % (top_gene, cell_type_label)\n",
    "        \n",
    "        for other_cluster in other_clusters:\n",
    "            new_cluster_cell_type_label[other_cluster] = \"%s- %s\" % (top_gene, cell_type_label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_cluster_cell_type_label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cluster, label in new_cluster_cell_type_label.items():\n",
    "    \n",
    "    if label == \"Cldn5+ Pericytes\":\n",
    "        label = \"Multiplets\"\n",
    "    elif label == \"Cldn5- Pericytes\":\n",
    "        label = \"Pericytes\"\n",
    "    \n",
    "    cluster_cell_type_label[cluster] = label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_cell_type_label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENES_OF_INTEREST = set()\n",
    "\n",
    "for cell_type in CELL_TYPE_MARKER_GENES:\n",
    "    for cell_subtype in CELL_TYPE_MARKER_GENES[cell_type]:\n",
    "        GENES_OF_INTEREST.update(CELL_TYPE_MARKER_GENES[cell_type][cell_subtype])\n",
    "        \n",
    "GENES_OF_INTEREST = [\"C1ql1\"]\n",
    "\n",
    "for gene in GENES_OF_INTEREST:\n",
    "\n",
    "    ensembl_id = adata.var.loc[adata.var['Gene Name']==gene].index[0]\n",
    "    normalized_gene_counts = numpy.array(normalized_gene_expression.loc[:, ensembl_id].values).reshape((-1,))\n",
    "\n",
    "    aavomics.plot_gene_expression(adata.obsm[TRANSFORM_TO_PLOT], numpy.log2(normalized_gene_counts), filename=os.path.join(\"out\", \"gene_expression_%s_normalized_non_neuronal.html\" % gene))\n",
    "\n",
    "    raw_gene_counts = numpy.array(adata[:, ensembl_id].X.todense()).reshape((-1,))\n",
    "\n",
    "    aavomics.plot_gene_expression(adata.obsm[TRANSFORM_TO_PLOT], numpy.log2(raw_gene_counts+1), filename=os.path.join(\"out\", \"gene_expression_%s_raw_non_neuronal.html\" % gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_categorical_type = pandas.CategoricalDtype(categories=list(set(cluster_cell_type_label.values())))\n",
    "adata.obs[SUBTYPE_TAXONOMY_NAME] = pandas.Series(dtype=cell_type_categorical_type)\n",
    "\n",
    "for cluster, cell_type_label in cluster_cell_type_label.items():\n",
    "    adata.obs[SUBTYPE_TAXONOMY_NAME].loc[adata.obs[CLUSTER_OBS_NAME] == cluster] = cell_type_label\n",
    "\n",
    "aavomics.plot_clusters(adata.obsm[TRANSFORM_TO_PLOT], adata.obs[SUBTYPE_TAXONOMY_NAME], filename=os.path.join(\"out\", \"cell_types_non_neuronal.html\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"p_%s\" % SUBTYPE_TAXONOMY_NAME] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "COLUMNS_TO_DROP = [SUBTYPE_TAXONOMY_NAME, \"p_%s\" % SUBTYPE_TAXONOMY_NAME]\n",
    "\n",
    "for cell_set_index, cell_set_name in enumerate(adata.obs[\"Cell Set\"].unique()):\n",
    "        \n",
    "    print(cell_set_name)\n",
    "    \n",
    "    cell_set = database.CELL_SETS_DICT[cell_set_name]\n",
    "    \n",
    "    anndata_file_path = cell_set.get_anndata_file_path(alignment_name=ALIGNMENT_NAME)\n",
    "    \n",
    "    if not os.path.exists(anndata_file_path):\n",
    "        print(\"Missing %s, skipping\" % cell_set.name)\n",
    "        continue\n",
    "    \n",
    "    cell_set_adata = anndata.read(anndata_file_path)\n",
    "    \n",
    "    for column in COLUMNS_TO_DROP:\n",
    "    \n",
    "        if column in cell_set_adata.obs.columns:\n",
    "            cell_set_adata.obs.drop(column, axis=1, inplace=True)\n",
    "        \n",
    "    adata_filtered = adata[adata.obs[\"Cell Set\"] == cell_set_name]\n",
    "    adata_filtered.obs.index = [\"-\".join(x.split(\"-\")[0:-1]) for x in adata_filtered.obs.index]\n",
    "    cell_set_adata.obs[SUBTYPE_TAXONOMY_NAME] = adata_filtered.obs[SUBTYPE_TAXONOMY_NAME]\n",
    "    cell_set_adata.obs[\"p_%s\" % SUBTYPE_TAXONOMY_NAME] = adata_filtered.obs[\"p_%s\" % SUBTYPE_TAXONOMY_NAME]\n",
    "    \n",
    "    cell_set_adata.write_h5ad(anndata_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
