{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import subprocess\n",
    "\n",
    "from aavomics import database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "BASH_SCRIPT_TEMPLATE = os.path.join(\"..\", \"bin\", \"cellranger_count.sh\")\n",
    "FILE_NAMES_TO_TRANSFER = [\"filtered_feature_bc_matrix.h5\", \"raw_feature_bc_matrix.h5\", \"possorted_genome_bam.bam\", \"possorted_genome_bam.bam.bai\", \"web_summary.html\"]\n",
    "FORCE_RERUN = False\n",
    "REMOVE_TMP_DIR = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for alignment in database.ALIGNMENTS:\n",
    "        \n",
    "    # Get the name of the reference we will use for this alignment\n",
    "    reference_name = database.ALIGNMENTS_DICT[alignment.name].reference.name\n",
    "\n",
    "    for cell_set in database.CELL_SETS:\n",
    "        \n",
    "        print(alignment.name, cell_set.name)\n",
    "\n",
    "        transcriptome_dir = os.path.join(database.DATA_PATH, \"cell_sets\", cell_set.name, \"transcriptome\")\n",
    "        alignment_dir = os.path.join(transcriptome_dir, \"transcripts\", alignment.name)\n",
    "        working_dir = os.path.join(alignment_dir, \"tmp\")\n",
    "            \n",
    "        # Check if the working directory already exists - this means an alignment may be in progress\n",
    "        if os.path.exists(working_dir):\n",
    "            if not REMOVE_TMP_DIR:\n",
    "                print(\"Skipping %s, alignment in progress detected\" % cell_set.name)\n",
    "                continue\n",
    "            else:\n",
    "                shutil.rmtree(working_dir)\n",
    "\n",
    "        # Check if this alignment has been completed already\n",
    "        already_exists = True\n",
    "\n",
    "        for file_name in FILE_NAMES_TO_TRANSFER:\n",
    "             if not os.path.exists(os.path.join(alignment_dir, file_name)):\n",
    "                    already_exists = False\n",
    "        \n",
    "        if already_exists:\n",
    "            if not FORCE_RERUN:\n",
    "                print(\"Skipping %s, alignment already detected\" % cell_set.name)\n",
    "                continue\n",
    "            else:\n",
    "                shutil.rmtree(alignment_dir)\n",
    "\n",
    "        if not cell_set.has_reads(read_type=\"Transcriptome\"):    \n",
    "            print(\"Skipping %s, doesn't have transcriptome reads\" % cell_set.name)\n",
    "            continue\n",
    "\n",
    "        read_set_names = set()\n",
    "\n",
    "        for sequencing_library in cell_set.sequencing_libraries:\n",
    "\n",
    "            if sequencing_library.type != \"Transcriptome\":\n",
    "                continue\n",
    "\n",
    "            for read_set in sequencing_library.read_sets:\n",
    "                read_set_names.add(read_set.name)\n",
    "\n",
    "        if len(read_set_names) == 0:\n",
    "            print(\"Skipping %s, can't find read sets\" % cell_set.name)\n",
    "            continue\n",
    "\n",
    "        if os.path.exists(alignment_dir):\n",
    "            shutil.rmtree(alignment_dir)\n",
    "\n",
    "        bash_script_file_name = \"%s.sh\" % cell_set.name\n",
    "        shutil.copy(BASH_SCRIPT_TEMPLATE, bash_script_file_name)\n",
    "\n",
    "        cellranger_args = {\n",
    "            \"id\": cell_set.name,\n",
    "            \"fastqs\": os.path.join(transcriptome_dir, \"reads\"),\n",
    "            \"sample\": \",\".join(read_set_names),\n",
    "            \"transcriptome\": os.path.join(database.DATA_PATH, \"references\", reference_name),\n",
    "            \"localcores\": \"32\"\n",
    "        }\n",
    "\n",
    "        with open(bash_script_file_name, \"a\") as bash_script_file:\n",
    "            \n",
    "            bash_script_file.write(\"mkdir -p %s\\n\" % working_dir)\n",
    "            bash_script_file.write(\"pushd %s\\n\" % working_dir)\n",
    "            bash_script_file.write(\"cellranger count\")\n",
    "\n",
    "            for key, value in cellranger_args.items():\n",
    "\n",
    "                if value is not None:\n",
    "                    bash_script_file.write(\" --%s=%s\" % (key, value))\n",
    "                else:\n",
    "                    bash_script_file.write(\" --%s\" % key)\n",
    "\n",
    "            bash_script_file.write(\"\\n\")\n",
    "            for file_name in FILE_NAMES_TO_TRANSFER:\n",
    "                bash_script_file.write(\"mv %s %s\\n\" % (os.path.join(working_dir, cell_set.name, \"outs\", file_name), alignment_dir))\n",
    "            bash_script_file.write(\"popd\\n\")\n",
    "            bash_script_file.write(\"rm -r %s\\n\" % working_dir)\n",
    "            bash_script_file.write(\"echo done\\n\")\n",
    "\n",
    "        print(\"Queueing %s\" % cell_set.name)\n",
    "        subprocess.run(\"sbatch %s\" % bash_script_file_name, shell=True)\n",
    "        os.remove(bash_script_file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
