{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from aavomics import database\n",
    "import os\n",
    "import shutil\n",
    "\n",
    "from pepars.utils import Illumina_FASTQ_File_Set\n",
    "from skbio.alignment import StripedSmithWaterman as SSW\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_cell_barcode_templates(FASTQ_file_set, template, cell_barcode_UMI_length):\n",
    "    \n",
    "    sequence_index = 0\n",
    "    alignment_counts = {}\n",
    "    sequence_alignments = {}\n",
    "    \n",
    "    cigar_regex = re.compile(r\"(\\d+)(\\w)\")\n",
    "    aligner = SSW(template, suppress_sequences=False)\n",
    "    \n",
    "    for sequences in FASTQ_file_set.get_sequence_iterator():\n",
    "        \n",
    "        read_1 = sequences[-2]\n",
    "        read_2 = sequences[-1]\n",
    "        \n",
    "        cell_barcode_UMI = read_1[0:cell_barcode_UMI_length]\n",
    "        \n",
    "        if read_2 in sequence_alignments:\n",
    "            alignment_tuple = sequence_alignments[read_2]\n",
    "        else:\n",
    "            alignment = aligner(read_2)\n",
    "            cigar_segments = cigar_regex.findall(alignment[\"cigar\"])\n",
    "\n",
    "            target_index = alignment[\"target_begin\"]\n",
    "            insertion_sequence = \"\"\n",
    "\n",
    "            for match_number, match_type in cigar_segments:\n",
    "\n",
    "                match_number = int(match_number)\n",
    "\n",
    "                if match_type == \"M\":\n",
    "                    target_index += match_number\n",
    "                elif match_type == \"I\":\n",
    "                    target_index -= match_number\n",
    "                elif match_type == \"D\":\n",
    "                    insertion_sequence += read_2[target_index:target_index+match_number]\n",
    "                    target_index += match_number\n",
    "                else:\n",
    "                    print(\"Match type is %s wut\" % match_type)\n",
    "            \n",
    "            alignment_max_score = min(len(template), len(read_2)) * 2\n",
    "            \n",
    "            alignment_score = int(alignment[\"optimal_alignment_score\"])            \n",
    "        \n",
    "            alignment_tuple = (alignment[\"cigar\"], alignment_score, alignment_max_score, int(alignment[\"target_begin\"]), int(alignment[\"query_begin\"]), insertion_sequence)\n",
    "        \n",
    "        cell_barcode_UMI_alignment_tuple = (cell_barcode_UMI, alignment_tuple)\n",
    "        \n",
    "        if cell_barcode_UMI_alignment_tuple not in alignment_counts:\n",
    "            alignment_counts[cell_barcode_UMI_alignment_tuple] = 1\n",
    "        else:\n",
    "            alignment_counts[cell_barcode_UMI_alignment_tuple] += 1\n",
    "            \n",
    "        sequence_index += 1\n",
    "        \n",
    "    return alignment_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cell_set in database.CELL_SETS:\n",
    "    \n",
    "    tissue_sample = cell_set.source_tissue\n",
    "    dissociation_run = tissue_sample.dissociation_run\n",
    "    animal = tissue_sample.animal\n",
    "    \n",
    "    if animal is None:\n",
    "        continue\n",
    "        \n",
    "    injections = animal.injections\n",
    "    \n",
    "    templates = set()\n",
    "    \n",
    "    for injection in injections:\n",
    "        for vector in injection.vector_pool.vectors:\n",
    "            templates.add(vector.cargo)\n",
    "            \n",
    "    if dissociation_run.protocol_version == 2:\n",
    "        cell_barcode_UMI_length = 26\n",
    "    else:\n",
    "        cell_barcode_UMI_length = 28\n",
    "    \n",
    "    if len(templates) == 0:\n",
    "        print(\"Skipping %s, no templates injected\" % cell_set.name)\n",
    "        continue\n",
    "    \n",
    "    amplified_directory = os.path.join(database.DATA_PATH, \"cell_sets\", cell_set.name, \"virus\", \"reads\")\n",
    "    \n",
    "    read_sets = set()\n",
    "    \n",
    "    for sequencing_library in cell_set.sequencing_libraries:\n",
    "        if sequencing_library.type == \"Virus Transcripts\":\n",
    "            read_sets.update(sequencing_library.read_sets)\n",
    "    \n",
    "    if len(read_sets) == 0:\n",
    "        print(\"Skipping %s, no amplified read sets detected\" % cell_set.name)\n",
    "        continue\n",
    "        \n",
    "    for template in templates:\n",
    "        \n",
    "        found_read_set = False\n",
    "        \n",
    "        for read_set in read_sets:\n",
    "                \n",
    "            alignment_file_name = \"%s_%s.csv\" % (read_set.name, template.name)\n",
    "            alignment_file_directory = os.path.join(database.DATA_PATH, \"cell_sets\", cell_set.name, \"virus\", \"alignments\")\n",
    "            alignment_file_path = os.path.join(alignment_file_directory, alignment_file_name)\n",
    "\n",
    "            if os.path.exists(alignment_file_path):\n",
    "                print(\"Skipping %s against %s, already exists\" % (read_set.name, template.name))\n",
    "                continue\n",
    "\n",
    "            template_sequence = template.sequence\n",
    "\n",
    "            print(\"Aligning %s against %s\" % (read_set.name, template.name))\n",
    "            \n",
    "            file_set = Illumina_FASTQ_File_Set(amplified_directory, read_set.name)\n",
    "\n",
    "            if len(file_set.file_paths) == 0:\n",
    "                print(\"Cant find %s in %s\" % (read_set.name, cell_set.name))\n",
    "                continue\n",
    "\n",
    "            alignments = align_cell_barcode_templates(file_set, template_sequence, cell_barcode_UMI_length)\n",
    "\n",
    "            if not os.path.exists(alignment_file_directory):\n",
    "                os.makedirs(alignment_file_directory)\n",
    "\n",
    "            temp_alignment_file_path = alignment_file_path + \".tmp\"\n",
    "\n",
    "            with open(temp_alignment_file_path, \"w\") as temp_alignment_file:\n",
    "                for alignment in sorted(alignments.items(), key=lambda x: x[1], reverse=True):\n",
    "                    temp_alignment_file.write(alignment[0][0])\n",
    "                    temp_alignment_file.write(\",\")\n",
    "                    temp_alignment_file.write(\",\".join([str(x) for x in alignment[0][1]]))\n",
    "                    temp_alignment_file.write(\",\")\n",
    "                    temp_alignment_file.write(str(alignment[1]))\n",
    "                    temp_alignment_file.write(\"\\n\")\n",
    "\n",
    "            shutil.move(src=temp_alignment_file_path, dst=alignment_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
