{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas\n",
    "\n",
    "from aavomics import database\n",
    "import anndata\n",
    "import pysam\n",
    "\n",
    "from skbio.alignment import StripedSmithWaterman as SSW"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALIGNMENT_NAME = \"cellranger_5.0.1_gex_mm10_2020_A_AAVomics\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats_dict = {}\n",
    "\n",
    "for cell_set in database.CELL_SETS:\n",
    "    \n",
    "    tissue_sample = cell_set.source_tissue\n",
    "    dissociation_run = tissue_sample.dissociation_run\n",
    "    animal = tissue_sample.animal\n",
    "    \n",
    "    if tissue_sample.region != \"Cortex\":\n",
    "        raise ValueError(\"Should only be cortex samples!\")\n",
    "    \n",
    "    if animal is None:\n",
    "        continue\n",
    "\n",
    "    if dissociation_run.protocol_version == 2:\n",
    "        cell_barcode_UMI_length = 26\n",
    "    else:\n",
    "        cell_barcode_UMI_length = 28\n",
    "\n",
    "    if animal.injections is None:\n",
    "        print(\"%s has no injections, skipping.\" % cell_set.name)\n",
    "        continue\n",
    "        \n",
    "    injections = animal.injections\n",
    "\n",
    "    # Get all the vectors associated with the same template\n",
    "    template_vectors = {}\n",
    "\n",
    "    for injection in injections:\n",
    "        for vector in injection.vector_pool.vectors:\n",
    "            if vector.cargo not in template_vectors:\n",
    "                template_vectors[vector.cargo] = [vector]\n",
    "            else:\n",
    "                template_vectors[vector.cargo].append(vector)\n",
    "    \n",
    "    if len(template_vectors) == 0:\n",
    "        print(\"%s has no delivered genetic cargo, skipping.\" % cell_set.name)\n",
    "        continue\n",
    "    \n",
    "    viruses = set([vector.delivery_vehicle for template, vectors in template_vectors.items() for vector in vectors])\n",
    "        \n",
    "    adata_file_path = cell_set.get_anndata_file_path(alignment_name=ALIGNMENT_NAME)\n",
    "    \n",
    "    if not os.path.exists(adata_file_path):\n",
    "        print(\"%s is missing anndata file expected at %s\" % (cell_set.name, adata_file_path))\n",
    "        continue\n",
    "    \n",
    "    print(\"Processing %s now\" % cell_set.name)\n",
    "        \n",
    "    adata = anndata.read_h5ad(adata_file_path)\n",
    "    \n",
    "    alignment_directory = os.path.join(database.DATA_PATH, \"cell_sets\", cell_set.name, \"virus\", \"alignments\")\n",
    "\n",
    "    read_sets = set()\n",
    "\n",
    "    for sequencing_library in cell_set.sequencing_libraries:\n",
    "        if sequencing_library.type == \"Virus Transcripts\":\n",
    "            read_sets.update(sequencing_library.read_sets)\n",
    "\n",
    "    template_cell_barcode_alignments = {}\n",
    "    template_num_reads = {}\n",
    "\n",
    "    for template in template_vectors:\n",
    "\n",
    "        vectors = template_vectors[template]\n",
    "        \n",
    "        has_barcodes = False\n",
    "\n",
    "        if len(vectors) > 1:\n",
    "            has_barcodes = True\n",
    "            vector_aligners = []\n",
    "\n",
    "            for vector in vectors:\n",
    "\n",
    "                if not vector.barcode:\n",
    "                    continue\n",
    "\n",
    "                vector_aligner = SSW(vector.barcode, suppress_sequences=False, score_only=True)\n",
    "\n",
    "                vector_aligners.append([vector, vector_aligner])\n",
    "\n",
    "                for i in range(1, len(vector.barcode) + 1):\n",
    "\n",
    "                    if template.sequence[int(vector.barcode_location)-i] == vector.barcode[-i]:\n",
    "                        permuted_barcode = vector.barcode[-i:] + vector.barcode[0:-i]\n",
    "                        vector_aligner = SSW(permuted_barcode, suppress_sequences=False, score_only=True)\n",
    "                        vector_aligners.append([vector, vector_aligner])\n",
    "                    else:\n",
    "                        break\n",
    "\n",
    "                for i in range(len(vector.barcode)):\n",
    "\n",
    "                    if template.sequence[int(vector.barcode_location)+i] == vector.barcode[i]:\n",
    "                        permuted_barcode = vector.barcode[i+1:] + vector.barcode[-(i+1):]\n",
    "                        vector_aligner = SSW(permuted_barcode, suppress_sequences=False, score_only=True)\n",
    "                        vector_aligners.append([vector, vector_aligner])\n",
    "                    else:\n",
    "                        break\n",
    "            barcode_length = len(vector.barcode)\n",
    "        else:\n",
    "            barcode_length = 0\n",
    "\n",
    "        template_cell_barcode_alignments[template] = {}\n",
    "        template_num_reads[template] = 0\n",
    "        \n",
    "        insertion_best_vectors = {}\n",
    "        insertion_counts = {}\n",
    "\n",
    "        for read_set in read_sets:\n",
    "\n",
    "            alignment_file_name = \"%s_%s.csv\" % (read_set.name, template.name)\n",
    "            alignment_file_path = os.path.join(alignment_directory, alignment_file_name)\n",
    "\n",
    "            i = 0\n",
    "            \n",
    "            print(\"Parsing alignment file %s\" % read_set.name)\n",
    "            \n",
    "            with open(alignment_file_path, \"r\") as alignment_file:\n",
    "\n",
    "                line = alignment_file.readline()\n",
    "\n",
    "                while line:\n",
    "                \n",
    "                    i += 1\n",
    "                    \n",
    "                    if i % 100000 == 0:\n",
    "                        print(i)\n",
    "\n",
    "                    line_values = line.split(\",\")\n",
    "                    cell_barcode_UMI = line_values[0]\n",
    "                    score = int(line_values[2])\n",
    "                    max_score = int(line_values[3])\n",
    "                    insertion = line_values[-2]\n",
    "                    \n",
    "                    if has_barcodes:\n",
    "                        max_score -= 3 + 4*len(insertion)\n",
    "                        \n",
    "                    count = int(line_values[-1])\n",
    "\n",
    "                    line = alignment_file.readline()\n",
    "                    \n",
    "                    if score/max_score < 0.9:\n",
    "                        continue\n",
    "\n",
    "                    template_num_reads[template] += count\n",
    "\n",
    "                    if len(vectors) > 1:\n",
    "\n",
    "                        if len(insertion) == 0:\n",
    "                            continue\n",
    "                            \n",
    "                        if insertion not in insertion_best_vectors:\n",
    "                            \n",
    "                            insertion_counts[insertion] = 0\n",
    "                            \n",
    "                            vector_scores = [\n",
    "                                (vector, vector_aligner(insertion)[\"optimal_alignment_score\"])\n",
    "                                for vector, vector_aligner in vector_aligners\n",
    "                            ]\n",
    "\n",
    "                            sorted_scores = sorted(vector_scores, key=lambda x: x[1], reverse=True)\n",
    "\n",
    "                            if sorted_scores[0][1] == sorted_scores[1][1]:\n",
    "                                insertion_best_vectors[insertion] = (None, None)\n",
    "                            elif sorted_scores[0][1]/(barcode_length*2) < 0.9:\n",
    "                                insertion_best_vectors[insertion] = (None, None)\n",
    "                            else:\n",
    "                                insertion_best_vectors[insertion] = (sorted_scores[0][0], sorted_scores[0][1])\n",
    "                                \n",
    "                        best_vector = insertion_best_vectors[insertion][0]\n",
    "                        best_score = insertion_best_vectors[insertion][1]\n",
    "                        insertion_counts[insertion] += count\n",
    "                        \n",
    "                        if best_vector is None:\n",
    "                            continue\n",
    "                        \n",
    "                        score += 3 + 2 * best_score\n",
    "                    else:\n",
    "                        best_vector = vectors[0]\n",
    "\n",
    "                    if cell_barcode_UMI not in template_cell_barcode_alignments[template]:\n",
    "                        template_cell_barcode_alignments[template][cell_barcode_UMI] = [(score, max_score, count, best_vector)]\n",
    "                    else:\n",
    "                        template_cell_barcode_alignments[template][cell_barcode_UMI].append((score, max_score, count, best_vector))\n",
    "    \n",
    "    total_vector_counts = {vector: 0 for vector in template_vectors[template]}\n",
    "    total_template_counts = {template: 0 for template in template_vectors}\n",
    "    vector_counts_array = []\n",
    "\n",
    "    for cell_barcode_UMI, counts in template_cell_barcode_alignments[template].items():\n",
    "\n",
    "        cell_barcode_UMI_vector_counts = {}\n",
    "\n",
    "        for score_tuple in template_cell_barcode_alignments[template][cell_barcode_UMI]:\n",
    "\n",
    "            vector = score_tuple[3]\n",
    "\n",
    "            total_vector_counts[vector] += score_tuple[2]\n",
    "            \n",
    "            for template in template_vectors:\n",
    "                if vector in template_vectors[template]:\n",
    "                    break\n",
    "\n",
    "            if vector not in cell_barcode_UMI_vector_counts:\n",
    "                cell_barcode_UMI_vector_counts[vector] = score_tuple[2]\n",
    "                total_template_counts\n",
    "            else:\n",
    "                cell_barcode_UMI_vector_counts[vector] += score_tuple[2]\n",
    "\n",
    "        for vector, count in cell_barcode_UMI_vector_counts.items():\n",
    "            vector_counts_array.append(count)\n",
    "\n",
    "    print(\"Done parsing alignment files\")\n",
    "            \n",
    "    BAM_file_path = os.path.join(\n",
    "            database.DATA_PATH, \"cell_sets\", cell_set.name, \"transcriptome\", \"transcripts\", ALIGNMENT_NAME, \"possorted_genome_bam.bam\")\n",
    "\n",
    "    samfile = pysam.AlignmentFile(BAM_file_path, \"rb\")\n",
    "\n",
    "    template_cell_barcode_UMI_counts = {}\n",
    "    template_cell_barcode_UMI_vectors = {}\n",
    "\n",
    "    for template in template_vectors:\n",
    "\n",
    "        template_cell_barcode_UMI_counts[template] = {}\n",
    "        template_cell_barcode_UMI_vectors[template] = {}\n",
    "        \n",
    "    vector_total_counts  = {}\n",
    "\n",
    "    for vector in vectors:\n",
    "\n",
    "        vector_total_counts[vector] = 0\n",
    "\n",
    "        for cell_barcode_UMI in template_cell_barcode_alignments[template]:\n",
    "\n",
    "            for score_tuple in template_cell_barcode_alignments[template][cell_barcode_UMI]:\n",
    "\n",
    "                if vector != score_tuple[3]:\n",
    "                    continue\n",
    "\n",
    "                vector_total_counts[vector] += score_tuple[2]\n",
    "\n",
    "    num_reads = 0\n",
    "    num_reads_passing_filter = 0\n",
    "    num_reads_matching_template = 0\n",
    "    num_reads_with_cell_barcode_UMI_in_amplified = 0\n",
    "    num_reads_unique_to_template = 0\n",
    "    num_reads_unique_vector = 0\n",
    "\n",
    "    for read in samfile.fetch(\"AAV\"):\n",
    "\n",
    "        num_reads += 1\n",
    "\n",
    "        if read.mapping_quality != 255:\n",
    "            continue\n",
    "\n",
    "        if not read.has_tag(\"CB\") or not read.has_tag(\"UB\"):\n",
    "            continue\n",
    "\n",
    "        filter_bits = read.get_tag(\"xf\")\n",
    "\n",
    "        if not (filter_bits & 1):\n",
    "            continue\n",
    "\n",
    "        if (filter_bits & 2):\n",
    "            continue\n",
    "\n",
    "        if (filter_bits & 32):\n",
    "            continue\n",
    "\n",
    "        if not read.has_tag(\"TX\"):\n",
    "            continue\n",
    "\n",
    "        num_reads_passing_filter += 1\n",
    "\n",
    "        matching_templates = set([x.split(\",\")[0] for x in read.get_tag(\"TX\").split(\";\")])\n",
    "\n",
    "        matches_template = False\n",
    "\n",
    "        for template in template_vectors:\n",
    "            if template.name in matching_templates:\n",
    "                matches_template = True\n",
    "                break\n",
    "\n",
    "        if not matches_template:\n",
    "            continue\n",
    "\n",
    "        num_reads_matching_template += 1\n",
    "\n",
    "        cell_barcode_UMI = read.get_tag(\"CB\")[0:16] + read.get_tag(\"UB\")\n",
    "\n",
    "        cell_barcode_UMI_exists_in_amplified = False\n",
    "\n",
    "        for template in template_vectors:\n",
    "            if cell_barcode_UMI in template_cell_barcode_alignments[template]:\n",
    "                cell_barcode_UMI_exists_in_amplified = True\n",
    "                break\n",
    "\n",
    "        if not cell_barcode_UMI_exists_in_amplified:\n",
    "            continue\n",
    "\n",
    "        num_reads_with_cell_barcode_UMI_in_amplified += 1\n",
    "\n",
    "        template_counts = {}\n",
    "\n",
    "        for template in template_vectors:\n",
    "\n",
    "            if cell_barcode_UMI in template_cell_barcode_alignments[template]:\n",
    "\n",
    "                cell_barcode_UMI_reads = sum(x[2] for x in template_cell_barcode_alignments[template][cell_barcode_UMI])\n",
    "                count = cell_barcode_UMI_reads/template_num_reads[template]\n",
    "\n",
    "                template_counts[template] = count\n",
    "\n",
    "        if len(template_counts) > 1:\n",
    "\n",
    "            sorted_template_counts = sorted(template_counts.items(), key=lambda x: x[1], reverse=True)\n",
    "\n",
    "            top_template_count = sorted_template_counts[0][1]\n",
    "            second_template_count = sorted_template_counts[1][1]\n",
    "\n",
    "            if top_template_count == second_template_count:\n",
    "                template = None\n",
    "            else:\n",
    "                template = sorted_template_counts[0][0]\n",
    "        else:\n",
    "            template = sorted(template_counts.items(), key=lambda x: x[1], reverse=True)[0][0]\n",
    "\n",
    "        if template is None:\n",
    "            continue\n",
    "\n",
    "        num_reads_unique_to_template += 1\n",
    "\n",
    "        if cell_barcode_UMI not in template_cell_barcode_UMI_counts[template]:\n",
    "\n",
    "            if len(template_cell_barcode_alignments[template][cell_barcode_UMI]) > 1:\n",
    "\n",
    "                vector_counts = {}\n",
    "\n",
    "                for score_tuple in template_cell_barcode_alignments[template][cell_barcode_UMI]:\n",
    "\n",
    "                    vector = score_tuple[3]\n",
    "\n",
    "                    if vector not in vector_counts:\n",
    "                        vector_counts[vector] = score_tuple[2]\n",
    "                    else:\n",
    "                        vector_counts[vector] += score_tuple[2]\n",
    "                        \n",
    "                if len(vector_counts) > 1:\n",
    "                    sorted_vector_counts = sorted(vector_counts.items(), key=lambda x: x[1], reverse=True)\n",
    "\n",
    "                    highest_count = sorted_vector_counts[0][1]\n",
    "                    highest_vector = sorted_vector_counts[0][0]\n",
    "\n",
    "                    second_highest_count = sorted_vector_counts[1][1]\n",
    "                    second_highest_vector = sorted_vector_counts[1][0]\n",
    "                    \n",
    "                    if highest_count == second_highest_count:\n",
    "                        continue\n",
    "                    else:\n",
    "                        highest_vector = sorted_vector_counts[0][0]\n",
    "                else:\n",
    "                    highest_vector = vector\n",
    "\n",
    "                highest_count = vector_counts[vector]\n",
    "            else:\n",
    "                highest_vector = template_cell_barcode_alignments[template][cell_barcode_UMI][0][3]\n",
    "                highest_count = template_cell_barcode_alignments[template][cell_barcode_UMI][0][2]\n",
    "\n",
    "            template_cell_barcode_UMI_vectors[template][cell_barcode_UMI] = highest_vector\n",
    "            template_cell_barcode_UMI_counts[template][cell_barcode_UMI] = 1\n",
    "            num_reads_unique_vector += 1\n",
    "        else:\n",
    "            template_cell_barcode_UMI_counts[template][cell_barcode_UMI] += 1\n",
    "            num_reads_unique_vector += 1\n",
    "            \n",
    "    stats_dict[cell_set.name] = {\n",
    "        \"num_reads\": num_reads,\n",
    "        \"num_reads_passing_filter\": num_reads_passing_filter,\n",
    "        \"num_reads_matching_template\": num_reads_matching_template,\n",
    "        \"num_reads_with_cell_barcode_UMI_in_amplified\": num_reads_with_cell_barcode_UMI_in_amplified,\n",
    "        \"num_reads_unique_to_template\": num_reads_unique_to_template,\n",
    "        \"num_reads_unique_vector\": num_reads_unique_vector\n",
    "    }\n",
    "                \n",
    "    for template, vectors in template_vectors.items():\n",
    "        \n",
    "        for vector in vectors:\n",
    "            \n",
    "            cell_barcode_counts = {x: 0 for x in adata.obs.index.values}\n",
    "            \n",
    "            for cell_barcode_UMI, virus_count in template_cell_barcode_UMI_counts[template].items():\n",
    "\n",
    "                if vector != template_cell_barcode_UMI_vectors[template][cell_barcode_UMI]:\n",
    "                    continue\n",
    "\n",
    "                cell_barcode = cell_barcode_UMI[0:16] + \"-1\"\n",
    "\n",
    "                if cell_barcode not in cell_barcode_counts:\n",
    "                    continue\n",
    "\n",
    "                cell_barcode_counts[cell_barcode] += 1\n",
    "\n",
    "            vector_counts_df = pandas.DataFrame.from_dict(cell_barcode_counts, orient=\"index\", columns=[template.name])\n",
    "            \n",
    "            if vector.name in adata.obs.columns:\n",
    "                adata.obs.drop(vector.name, axis=1, inplace=True)\n",
    "                \n",
    "            adata.obs[vector.name] = vector_counts_df\n",
    "\n",
    "    for virus in viruses:\n",
    "\n",
    "        cell_barcode_counts = {x: 0 for x in adata.obs.index.values}\n",
    "\n",
    "        for template, vectors in template_vectors.items():\n",
    "\n",
    "            for vector in vectors:\n",
    "\n",
    "                if vector.delivery_vehicle != virus:\n",
    "                    continue\n",
    "\n",
    "                for cell_barcode_UMI, virus_count in template_cell_barcode_UMI_counts[template].items():\n",
    "\n",
    "                    if vector != template_cell_barcode_UMI_vectors[template][cell_barcode_UMI]:\n",
    "                        continue\n",
    "\n",
    "                    cell_barcode = cell_barcode_UMI[0:16] + \"-1\"\n",
    "\n",
    "                    if cell_barcode not in cell_barcode_counts:\n",
    "                        continue\n",
    "\n",
    "                    cell_barcode_counts[cell_barcode] += 1\n",
    "\n",
    "        virus_counts_df = pandas.DataFrame.from_dict(cell_barcode_counts, orient=\"index\", columns=[template.name])\n",
    "            \n",
    "        if virus.name in adata.obs.columns:\n",
    "            adata.obs.drop(virus.name, axis=1, inplace=True)\n",
    "        \n",
    "        adata.obs[virus.name] = virus_counts_df\n",
    "    \n",
    "    adata.write_h5ad(adata_file_path)\n",
    "    \n",
    "    display(adata.obs.sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
